### ISSUE TITLE
Bug: [no-unnecessary-condition] false positives on type predicates with generic types

### ISSUE THREAD

**Issue by andreww2012** [typescript-eslint/typescript-eslint#10142]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=PTAEDECcHsFtQBYBckAcDOAuEBzAlkggK4BGAdAMZzDp4B2AJpAKbrSTHrBICeqzAWgBmrJMBIAbaCWAAJAKIBBACI1oRSBWY1mSAXWh19RCRICGk5mQZkk6ALAAoZgA9U7JKF79QAZV0AcoYBJuaWADwAQmbozAAqfMwANKAA0sw86KCuSMyMWQDWGdBCoNGxCT4AvKBFPCVlMfGJAHygNQDeTqCgANrpPKD0tcWl5c38ALqYaRnZLrn5s5ndPaAA-KBBdCGmFhLMUU2VzP0Zky2rPTPjJ2c8kwDcTgC%2Bz45Oru6QnlR06J4DDtQnh0Ah2qBwnEWgAKABuZgkRGYMzioAAPqA6KEMaAiIxmEJ6MwGABKGYIpHMIZZNFVNqU5GgACENWxpnenzcHlAfwBiBi212ElBCHAeGYEgYEKhKVS80WDEKo1A0Jhq2kACtUUlVnUZqldY5yaAtTS-IFgqF9oc4nK2vSsVbTKKYVrenVJqTHqAQJCBAInEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2pETRoHaJAA0mKGQAW-ANYAVEogAK0RDLJD8KRfjiIwAXwC64DyHdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// From https://github.com/sindresorhus/type-fest/blob/HEAD/source/set-non-nullable.d.ts
export type SetNonNullable<BaseType, Keys extends keyof BaseType = keyof BaseType> = {
  [Key in keyof BaseType]: Key extends Keys
    ? NonNullable<BaseType[Key]>
    : BaseType[Key];
};

export const nonNullish = <T>(value: T | null | undefined): value is T => value != null;

export const hasNonNullishField = <T, K extends keyof T>(
  obj: T,
  key: K,
): obj is SetNonNullable<T, K> => nonNullish(obj[key]); // <--
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error", { "checkTypePredicates": true }]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

The type of `obj[key]` in the example can be nullish and therefore the error should not be reported.

### Actual Result

The error is reported.

**Comment by kirkwaiblinger** [comment]
Hi @andreww2012 

This is an interesting example!

Let's look at a simplified repro:
[playground](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=MYewdgzgLgBG4DkCuAbFBLCALGBeGAPACoB8AFAG4CGKSApgFwxEwA%2BMSYAJnQGbpg6XNnFQoAlE2q06MTMzwkY0%2BjACE%2BMGIDcAWABQB3p2BR04GAHM6ggE7pgxcgA8mRSTGdyIMBIjFUAEYodE4wAN4GMDC2dFBItmBw-miYWGTO4gYAvkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2pADuIppAA0mKGQAW-ANYAVEogAK0RDLJD8KRfjiIwAXwC64dyDdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
```ts
const nonNullish = <T>(value: T | undefined | null): value is T => value != null;

function generic<T>(x: T): x is NonNullable<T> {
  return nonNullish(x)
}
```

The rule is right to report here - `x` has type `T` going into the `nonNullish` check and afterwards.

The issue is that the signature of `nonNullish` has a bug. You can trick it into believing false things like so:

```ts
const nonNullish = <T>(value: T | undefined | null): value is T => value != null;
declare const s: string | undefined | null;
if (nonNullish<string | undefined>(s)) {
     // TS Error: s is possibly undefined.
     s.toLowerCase();
}
```

Instead, you want to use 
```ts
const nonNullish = <T>(value: T): value is NonNullable<T> => value != null;
```

And that will make the lint error go away ([playground of your example with this change](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=PTAEDECcHsFtQBYBckAcDOAuEBzAlkggK4BGAdAMZzDp4B2AJpAKbrSTHrBICeqzAWgBmrJMBIAbaCWAAJAKIBBACI1oRSBWY1mSAXWh19RCRICGk5mQZkk6ALAAoZgA9U7JKF79QAZV0AcoYBJuaWADwAQmbozAAqfMwANKAA0sw86KCuSMyMWQDWGdBCoNGxCT4AvKBFPCVlMfGJAHygNQDeTqCgANrpPKD0tcWl5c38ALqYaRnZLrn5s5ndPaAA-KBBdCGmFhLMUU2VzP0Zky2rPTPjJ2c8kwDcTgC%2Bz45Oru6QnlR06J4DDtQnh0Ah2qBwnEWgAKABuZgkRGYMziAEoZgikcwhlltrswgcoW0qm0scjQABCGp0ULvT5uDygP4AxAxfEgsHgPDMCQMCFQlKpeaLBiFUagaEw1bSABWqKSqzqM1SiscGNActxfkCwVC%2B0OcSFJLaQIJoIQMLlvTqkzRj1AIEhAgETiAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2pETRoHaJAA0mKGQAW-ANYAVEogAK0RDLJD8KRfjiIwAXwC64DyHdA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

I _think_ that's the only issue here, not that there is a rule bug, but feel free to continue the conversation or file a new issue if you find another example.

**Comment by kirkwaiblinger** [comment]
side note - that faulty implementation of `nonNullish` is a pet peeve of mine that I've seen many times in the wild, including [(a variant thereof) in our repo](https://github.com/typescript-eslint/typescript-eslint/pull/8216/files#diff-b5f07d835e0413b96fa04f3a563a23bef62e787009ce4b482a34bc2d936146d9R14)!

**Comment by andreww2012** [comment]
Hello and thank you for the detailed analysis!

Now I see the issue with my `nonNullish` signature - it doesn't enforce `T` to be non-nullish which may happen if a type parameter is explicitly specified.

Turned out this rule has actually found a bug (in a peculiar way) as well as a few type predicates that shouldn't really be type predicates. So it worked flawlessly on a relatively big project :)

**Comment by kirkwaiblinger** [comment]
@andreww2012 thank you so much for the kind words 🙂 and I am quite glad to hear that this feature has been beneficial to you!
